Output output

wangxiaoxuan273 opened the pull request:
**What this PR does / why we need it**:
Implements output handlers for `manifest index` commands.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1503 

**Please check the following list**:
- [ ]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

<!-- If this is a security issue, please do not discuss on GitHub. Please report any suspected or confirmed security issues directly to https://github.com/oras-project/oras/blob/main/OWNERS.md. -->

qweeah commented on cmd/oras/internal/display/content/manifest_index.go:
```diff
@@ -0,0 +1,54 @@
+		defer f.Close()
+		out = f
+	}
+	return output.PrintJSON(out, manifest, h.pretty)
```
Should reset `pretty` if it's not output to stdout.

wangxiaoxuan273 replied:
Reset pretty is done in NewHandler.

qweeah commented on cmd/oras/internal/display/status/interface.go:
```diff
@@ -61,3 +63,24 @@ type CopyHandler interface {
+type ManifestIndexCreateHandler interface {
+	OnSourceManifestFetching(source string) error
+	OnSourceManifestFetched(source string) error
+	OnIndexPacked(shortDigest string) error
```
Packed index should be moved to metadata output.

wangxiaoxuan273 replied:
Packed, Updated, Pushed, Added, Removed, Merged should all be moved to metadata output.

wangxiaoxuan273 replied:
Use one interface for packed and updated.

wangxiaoxuan273 replied:
Created `manifestPackHandler` interface that contains `OnIndexPacked`, `OnIndexPushed` and `OnCompleted`. Removed `OnIndexUpdate` and replaced it with `OnIndexPacked`.

qweeah commented on cmd/oras/internal/display/status/interface.go:
```diff
@@ -61,3 +63,24 @@ type CopyHandler interface {
+
+// ManifestIndexCreateHandler handles status output for manifest index create command.
+type ManifestIndexCreateHandler interface {
+	OnSourceManifestFetching(source string) error
+	OnSourceManifestFetched(source string) error
```
Can we remove `Source` from the naming? Will non-source manifest be fetched?

wangxiaoxuan273 replied:
This is due to discard implementation has two OnManifestFetching method. To solve this issue,  an interface of ManifestReferenceFetchHandler is created and both index create and index update handlers contain it.

qweeah commented on cmd/oras/internal/display/status/interface.go:
```diff
@@ -61,3 +63,24 @@ type CopyHandler interface {
+	OnIndexFetched(indexRef string, digest digest.Digest) error
+	OnManifestFetching(manifestRef string) error
+	OnManifestFetched(manifestRef string, digest digest.Digest) error
+	OnManifestRemoved(digest digest.Digest) error
+	OnManifestAdded(manifestRef string, digest digest.Digest) error
+	OnIndexMerged(indexRef string, digest digest.Digest) error
```
Should be moved to metadata handler

wangxiaoxuan273 replied:
Moved Packed, Updated, Pushed, Added, Removed, Merged to metadata handlers.

qweeah commented on cmd/oras/internal/display/status/interface.go:
```diff
@@ -61,3 +63,24 @@ type CopyHandler interface {
+}
+
+// ManifestIndexUpdateHandler handles status output for manifest index update command.
+type ManifestIndexUpdateHandler interface {
```
Contains duplicated methods `ManifestIndexCreateHandler` as below

```golang
	OnIndexFetching(indexRef string) error
	OnIndexFetched(indexRef string, digest digest.Digest) error
```

You can create a new interface and use it in both interfaces.

wangxiaoxuan273 replied:
Created an interface `ManifestReferenceFetchHandler` and both index create and index update handlers contain it.

wangxiaoxuan273 replied:
Based on this [comment](https://github.com/oras-project/oras/pull/1509#discussion_r1796515589), this interface is now removed.

qweeah commented on cmd/oras/internal/display/handler.go:
```diff
@@ -174,9 +174,56 @@ func NewManifestPushHandler(printer *output.Printer) metadata.ManifestPushHandle
+	metadata.ManifestIndexCreateHandler,
+	content.ManifestIndexCreateHandler,
+	error) {
+	var statusHandler status.ManifestIndexCreateHandler
+	var metadataHandler metadata.ManifestIndexCreateHandler
+	var contentHandler content.ManifestIndexCreateHandler
+	switch outputPath {
+	case "":
+		statusHandler = status.NewTextManifestIndexCreateHandler(printer)
+		metadataHandler = text.NewManifestIndexCreateHandler(printer)
+		contentHandler = content.NewDiscardHandler()
+	case "-":
+		statusHandler = status.NewDiscardHandler()
+		metadataHandler = metadata.NewDiscardHandler()
+		contentHandler = content.NewManifestIndexCreateHandler(printer, pretty, outputPath)
+	default:
+		statusHandler = status.NewTextManifestIndexCreateHandler(printer)
+		metadataHandler = text.NewManifestIndexCreateHandler(printer)
+		contentHandler = content.NewManifestIndexCreateHandler(printer, pretty, outputPath)
+	}
```
```suggestion
	statusHandler := status.NewTextManifestIndexCreateHandler(printer)
	metadataHandler := text.NewManifestIndexCreateHandler(printer)
	contentHandler := content.NewManifestIndexCreateHandler(printer, pretty, outputPath)
	switch outputPath {
	case "":
		contentHandler = content.NewDiscardHandler()
	case "-":
		statusHandler = status.NewDiscardHandler()
		metadataHandler = metadata.NewDiscardHandler()
	}
```

qweeah commented on cmd/oras/root/manifest/index/update.go:
```diff
@@ -80,6 +79,10 @@ Example - Update an index and output the index to stdout, auto push will be disa
 				return err
 			}
 			opts.RawReference = args[0]
+			// ignore --pretty when output to a file
+			if opts.outputPath != "" && opts.outputPath != "-" {
+				opts.Pretty.Pretty = false
+			}
```
Can be merged into the handler construction.

TerryHowe replied:
If this is a general option thing, maybe a method in otpions?

wangxiaoxuan273 replied:
> If this is a general option thing, maybe a method in otpions?

@TerryHowe We can move it to content handler to avoid code repeat.

TerryHowe commented on cmd/oras/internal/display/status/text.go:
```diff
@@ -179,3 +181,106 @@ func (ch *TextCopyHandler) OnMounted(_ context.Context, desc ocispec.Descriptor)
+
+// NewTextManifestIndexUpdateHandler returns a new handler for manifest index create command.
+func NewTextManifestIndexUpdateHandler(printer *output.Printer) ManifestIndexUpdateHandler {
+	miuh := TextManifestIndexUpdateHandler{
+		printer: printer,
+	}
+	return &miuh
```
```suggestion
	manifestHandler := TextManifestIndexUpdateHandler{
		printer: printer,
	}
	return &manifestHandler
```

wangxiaoxuan273 replied:
The current naming style is aligned with the rest of the file.

TerryHowe commented on cmd/oras/root/manifest/index/create.go:
```diff
@@ -91,6 +89,10 @@ Example - Create an index and output the index to stdout, auto push will be disa
 			opts.sources = args[1:]
+			// ignore --pretty when output to a file
+			if opts.outputPath != "" && opts.outputPath != "-" {
+				opts.Pretty.Pretty = false
```
In the option handling in command maybe?

wangxiaoxuan273 replied:
> In the option handling in command maybe?

It's better to put the reset in content handler, since both `index create` and `index update` can use this code snippet from the handler. @TerryHowe

wangxiaoxuan273 replied:
Moved this part to `content.NewManifestIndexCreateHandler`

TerryHowe commented on cmd/oras/root/manifest/index/create.go:
```diff
@@ -188,7 +197,8 @@ func getPlatform(ctx context.Context, target oras.ReadOnlyTarget, manifestBytes
 }
 
-func pushIndex(ctx context.Context, target oras.Target, desc ocispec.Descriptor, content []byte, ref string, extraRefs []string, path string, printer *output.Printer) error {
+func pushIndex(ctx context.Context, onIndexPushed func(path string) error, onTagged func(desc ocispec.Descriptor, tag string) error,
```
This is a private method here, seems like it would make more sense for it to take handlers than function pointers.

Another thing, it seems like an extremely long list of params like something is not right here.

wangxiaoxuan273 replied:
Since both `create` and `update` use this method and index create handlers and index update handlers have different types, the function argument can't take handler argument.

Yeah this is an extremely long list of parameters, but I can't find another way :(

wangxiaoxuan273 commented on cmd/oras/internal/display/status/text.go:
```diff
@@ -179,3 +181,106 @@ func (ch *TextCopyHandler) OnMounted(_ context.Context, desc ocispec.Descriptor)
+
+// NewTextManifestIndexCreateHandler returns a new handler for manifest index create command.
+func NewTextManifestIndexCreateHandler(printer *output.Printer) ManifestIndexCreateHandler {
+	tmich := TextManifestIndexCreateHandler{
+		printer: printer,
+	}
+	return &tmich
```
The current naming style is aligned with the rest of the file.

wangxiaoxuan273 commented on cmd/oras/internal/display/content/manifest_index.go:
```diff
@@ -0,0 +1,54 @@
+// OnContentCreated is called after index content is created.
+func (h *manifestIndexCreate) OnContentCreated(manifest []byte) error {
+	out := h.stdout
+	if h.outputPath != "-" && h.outputPath != "" {
```
Did you mean to put "" before "-"? Changed the if condition to `if outputPath != "" && outputPath != "-"`

wangxiaoxuan273 commented on cmd/oras/internal/display/handler.go:
```diff
@@ -174,9 +174,56 @@ func NewManifestPushHandler(printer *output.Printer) metadata.ManifestPushHandle
+}
+
+// NewManifestIndexUpdateHandler returns status, metadata and content handlers for index update command.
+func NewManifestIndexUpdateHandler(outputPath string, printer *output.Printer, pretty bool) (
```
Currently we need both create and update handlers.

wangxiaoxuan273 commented on cmd/oras/internal/display/metadata/text/manifest_index.go:
```diff
@@ -0,0 +1,109 @@
+	return h.printer.Println("Digest:", desc.Digest)
+}
+
+type ManifestIndexUpdateHandler struct {
```
Need doc comment here.

wangxiaoxuan273 commented on cmd/oras/internal/display/metadata/text/manifest_index.go:
```diff
@@ -0,0 +1,109 @@
+	return h.printer.Println("Packed", descriptor.ShortDigest(desc), ocispec.MediaTypeImageIndex)
+}
+
+// OnIndexPushed implements metadata.ManifestIndexCreateHandler.
```
nit: remove metadata. here in the doc comment

wangxiaoxuan273 commented on cmd/oras/root/manifest/index/create.go:
```diff
@@ -199,13 +205,14 @@ func pushIndex(ctx context.Context, target oras.Target, desc ocispec.Descriptor,
-		handler := display.NewManifestIndexCreateHandler(printer)
-		tagListener := listener.NewTaggedListener(target, handler.OnTagged)
+		tagListener := listener.NewTaggedListener(target, onTagged)
 		if _, err = oras.TagBytesN(ctx, tagListener, desc.MediaType, content, extraRefs, oras.DefaultTagBytesNOptions); err != nil {
```
Just return the function result instead of error check.

wangxiaoxuan273 commented on cmd/oras/root/manifest/index/update.go:
```diff
@@ -114,116 +114,134 @@ func updateIndex(cmd *cobra.Command, opts updateOptions) error {
-func fetchIndex(ctx context.Context, target oras.ReadOnlyTarget, opts updateOptions) (ocispec.Index, error) {
-	printUpdateStatus(status.IndexPromptFetching, opts.Reference, "", opts.Printer)
-	desc, content, err := oras.FetchBytes(ctx, target, opts.Reference, oras.DefaultFetchBytesOptions)
+func fetchIndex(ctx context.Context, handler status.ManifestIndexUpdateHandler, target oras.ReadOnlyTarget, reference string) (ocispec.Index, error) {
```
rename the variable handler.

wangxiaoxuan273 commented on cmd/oras/root/manifest/index/update.go:
```diff
@@ -114,116 +114,134 @@ func updateIndex(cmd *cobra.Command, opts updateOptions) error {
 }
 
-func removeManifests(ctx context.Context, manifests []ocispec.Descriptor, target oras.ReadOnlyTarget, opts updateOptions) ([]ocispec.Descriptor, error) {
+func removeManifests(handler metadata.ManifestIndexUpdateHandler, manifests []ocispec.Descriptor, target oras.ReadOnlyTarget, opts updateOptions) ([]ocispec.Descriptor, error) {
```
change the opts argument to only opts.Reference.

qweeah commented on cmd/oras/internal/display/metadata/text/manifest_index.go:
```diff
@@ -0,0 +1,110 @@
+
+// ManifestIndexUpdateHandler handles text metadata output for index update events.
+type ManifestIndexUpdateHandler struct {
+	printer *output.Printer
```
Can you help separate `ManifestIndexUpdateHandler` and `ManifestIndexCreateHandler` into `manifest_index_update.go` and `manifest_index_create.go`?

qweeah commented on cmd/oras/internal/display/metadata/interface.go:
```diff
@@ -78,9 +79,25 @@ type ManifestPushHandler interface {
+type ManifestIndexUpdateHandler interface {
+	TaggedHandler
+	manifestPackHandler
+	OnManifestRemoved(digest digest.Digest) error
+	OnManifestAdded(manifestRef string, digest digest.Digest) error
+	OnIndexMerged(indexRef string, digest digest.Digest) error
 }
```
Change digest to descriptor since it will be used for formatted output in the future.

wangxiaoxuan273 replied:
The parameter of `OnManifestRemoved(digest digest.Digest) error` must be a digest, since the descriptor is unknown. Changed the parameter of the other two functions as descriptors.

qweeah commented on cmd/oras/internal/display/handler.go:
```diff
@@ -174,9 +174,42 @@ func NewManifestPushHandler(printer *output.Printer) metadata.ManifestPushHandle
+	status.ManifestIndexCreateHandler,
+	metadata.ManifestIndexCreateHandler,
+	content.ManifestIndexCreateHandler,
+	error) {
```
When will an error be returned?

wangxiaoxuan273 replied:
good catch. Removed the error return.

qweeah commented on cmd/oras/internal/display/status/utils.go:
```diff
@@ -52,12 +52,6 @@ const (
 const (
 	IndexPromptFetching = "Fetching "
```
Extra space, same for fetched.

```suggestion
	IndexPromptFetching = "Fetching"
```

wangxiaoxuan273 replied:
We are not doing status output indentation align anymore?

qweeah commented on cmd/oras/root/manifest/index/create.go:
```diff
@@ -188,7 +193,8 @@ func getPlatform(ctx context.Context, target oras.ReadOnlyTarget, manifestBytes
 }
 
-func pushIndex(ctx context.Context, target oras.Target, desc ocispec.Descriptor, content []byte, ref string, extraRefs []string, path string, printer *output.Printer) error {
+func pushIndex(ctx context.Context, onIndexPushed func(path string) error, onTagged func(desc ocispec.Descriptor, tag string) error,
+	target oras.Target, desc ocispec.Descriptor, content []byte, ref string, extraRefs []string, path string) error {
```
Try using handlers like:
```suggestion
func pushIndex(ctx context.Context, pmh metadata.ManifestPackHandler, th metadata.TaggedHandler,target oras.Target, desc ocispec.Descriptor, content []byte, ref string, extraRefs []string, path string) error {
```

wangxiaoxuan273 replied:
This function signature is now `func pushIndex(ctx context.Context, displayStatus status.ManifestIndexCreateHandler, taggedHandler metadata.TaggedHandler, target oras.Target, desc ocispec.Descriptor, content []byte, ref string, extraRefs []string, path string)` after resolving this [comment](https://github.com/oras-project/oras/pull/1509#discussion_r1796514181). Still long param list.

qweeah commented on cmd/oras/internal/display/content/manifest_index.go:
```diff
@@ -0,0 +1,58 @@
+
+// NewManifestIndexCreateHandler creates a new handler.
+func NewManifestIndexCreateHandler(out io.Writer, pretty bool, outputPath string) ManifestIndexCreateHandler {
+	// ignore --pretty when output to a file
+	if outputPath != "" && outputPath != "-" {
+		pretty = false
+	}
```
`NewManifestFetchHandler` is used only once by `oras manifest fetch` and such code snippet is in command code
https://github.com/oras-project/oras/blob/dcef719e208a9b226b15bc6512ad729a7dd93270/cmd/oras/root/manifest/fetch.go#L80-L83

I think we need to remove it and change the flag reset to `NewManifestFetchHandler` too.

wangxiaoxuan273 replied:
I can move the pretty reset of `oras manifest fetch` into the handler in a future PR, as this one is big enough.

wangxiaoxuan273 replied:
Move it in this PR.

wangxiaoxuan273 commented on cmd/oras/internal/display/content/discard.go:
```diff
@@ -24,7 +24,12 @@ func (discardHandler) OnContentFetched(ocispec.Descriptor, []byte) error {
+
 // NewDiscardHandler returns a new discard handler.
-func NewDiscardHandler() ManifestFetchHandler {
+func NewDiscardHandler() discardHandler {
```
Why is the export needed? It looks like even `status.DiscardHandler` exporting is not necessary.

wangxiaoxuan273 replied:
As `NewDiscardHandler` is a public method, its returned type should be exported.

wangxiaoxuan273 commented on cmd/oras/internal/display/content/manifest_index.go:
```diff
@@ -0,0 +1,58 @@
+		if err != nil {
+			return fmt.Errorf("failed to open %q: %w", h.outputPath, err)
+		}
+		defer f.Close()
```
Created #1516

wangxiaoxuan273 commented on cmd/oras/internal/display/metadata/discard.go:
```diff
@@ -15,7 +15,10 @@ limitations under the License.
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
 
 type discard struct{}
```
Is exporting necessary?

wangxiaoxuan273 replied:
exported

wangxiaoxuan273 commented on cmd/oras/internal/display/status/utils.go:
```diff
@@ -52,12 +52,6 @@ const (
 const (
 	IndexPromptFetching = "Fetching "
 	IndexPromptFetched  = "Fetched  "
```
The others are metadata prompts, and they have been moved to metadata handlers.

wangxiaoxuan273 replied:
As per offline discussion, added back "added, removed, merged, pushed, packed"

wangxiaoxuan273 commented:
As per offline discussion, added, merged, removed, pushed, packed should exist in both status and metadata output. And Status output is needed in the PR, metadata output can be added in a future pr when needed.

wangxiaoxuan273 commented on cmd/oras/internal/display/status/interface.go:
```diff
@@ -61,3 +62,17 @@ type CopyHandler interface {
+type referenceFetchHandler interface {
+	OnFetching(manifestRef string) error
+	OnFetched(manifestRef string, desc ocispec.Descriptor) error
+}
```
Removed this interface and now moved these three methods to index create handler.

shizhMSFT commented on cmd/oras/internal/display/content/manifest_index.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// OnContentCreated is called after index content is created.
+func (h *manifestIndexCreate) OnContentCreated(manifest []byte) error {
```
This method should be put after the constructor.

wangxiaoxuan273 replied:
moved.

shizhMSFT commented on cmd/oras/internal/display/handler.go:
```diff
@@ -174,9 +174,40 @@ func NewManifestPushHandler(printer *output.Printer) metadata.ManifestPushHandle
-func NewManifestIndexCreateHandler(printer *output.Printer) metadata.ManifestIndexCreateHandler {
-	return text.NewManifestIndexCreateHandler(printer)
+// NewManifestIndexCreateHandler returns status, metadata and content handlers for index create command.
+func NewManifestIndexCreateHandler(outputPath string, printer *output.Printer, pretty bool) (
+	status.ManifestIndexCreateHandler,
+	metadata.ManifestIndexCreateHandler,
+	content.ManifestIndexCreateHandler) {
```
nit: Could you make it one line?

shizhMSFT commented on cmd/oras/internal/display/handler.go:
```diff
@@ -174,9 +174,40 @@ func NewManifestPushHandler(printer *output.Printer) metadata.ManifestPushHandle
+	status.ManifestIndexCreateHandler,
+	metadata.ManifestIndexCreateHandler,
+	content.ManifestIndexCreateHandler) {
+	statusHandler := status.NewTextManifestIndexCreateHandler(printer)
+	metadataHandler := text.NewManifestIndexCreateHandler(printer)
+	contentHandler := content.NewManifestIndexCreateHandler(printer, pretty, outputPath)
+	switch outputPath {
+	case "":
+		contentHandler = content.NewDiscardHandler()
+	case "-":
+		statusHandler = status.NewDiscardHandler()
+		metadataHandler = metadata.NewDiscardHandler()
+	}
```
nit: Having initial values then overwritten is logically correct. However, it may cause unnecessary resource allocation. You may refactor them and make it clearer.

wangxiaoxuan273 replied:
Reverted this part back to the original version, before resolving this [comment](https://github.com/oras-project/oras/pull/1509#discussion_r1776443538)

shizhMSFT commented on cmd/oras/internal/display/status/text.go:
```diff
@@ -179,3 +182,93 @@ func (ch *TextCopyHandler) OnMounted(_ context.Context, desc ocispec.Descriptor)
+}
+
+// NewTextManifestIndexCreateHandler returns a new handler for manifest index create command.
+func NewTextManifestIndexCreateHandler(printer *output.Printer) ManifestIndexCreateHandler {
```
The correct order is 
1. struct definition
2. constructors
3. methods

shizhMSFT commented on cmd/oras/internal/display/status/text.go:
```diff
@@ -179,3 +182,93 @@ func (ch *TextCopyHandler) OnMounted(_ context.Context, desc ocispec.Descriptor)
+}
+
+// OnFetching implements ManifestIndexCreateHandler.
+func (mich TextManifestIndexCreateHandler) OnFetching(source string) error {
```
Should it be pointer receiver instead of value receiver?

shizhMSFT commented on cmd/oras/internal/display/status/text.go:
```diff
@@ -179,3 +182,93 @@ func (ch *TextCopyHandler) OnMounted(_ context.Context, desc ocispec.Descriptor)
+// OnManifestAdded implements metadata.ManifestIndexUpdateHandler.
+func (miuh TextManifestIndexUpdateHandler) OnManifestAdded(ref string, desc ocispec.Descriptor) error {
+	if contentutil.IsDigest(ref) {
+		return miuh.printer.Println("Added", ref)
```
Should we make those strings like `Added`, `Merged` constant?

qweeah commented on cmd/oras/root/manifest/index/update_test.go:
```diff
@@ -16,15 +16,96 @@ limitations under the License.
+	return nil
+}
+
+func NewTestUpdateDisplayStatus(onFetching, onFetched, onIndexPacked, onIndexPushed, onManifestRemoved, onManifestAdded, onIndexMerged bool) status.ManifestIndexUpdateHandler {
```
This function is hard to maintain since it is not friendly to future changes. E.g. whenever we need to add a new handler test, we need to modify ALL the existing usage of this function in ALL the test cases.

How about assembling the structure directly in the case array?

E.g.

For example:
```diff
{
    name:          "remove one matched item",
    manifests:     []ocispec.Descriptor{A, B, C},
    digestSet:     map[digest.Digest]bool{B.Digest: false},
-	displayStatus: NewTestUpdateDisplayStatus(false, false, false, false, false, false, false),
+	displayStatus: &testUpdateDisplayStatus{},
    indexRef:      "test01",
    want:          []ocispec.Descriptor{A, C},
    wantErr:       false,
}
```

Or

```diff
{
    name:          "handler error",
    manifests:     []ocispec.Descriptor{A, B, C},
    digestSet:     map[digest.Digest]bool{B.Digest: false},
-	displayStatus: NewTestUpdateDisplayStatus(false, false, false, false, true, false, false),
+	displayStatus: &testUpdateDisplayStatus{onManifestRemovedError: true},
    indexRef:      "test07",
    want:          nil,
    wantErr:       true,
},
```

qweeah replied:
Of course you can further apply a decorator pattern like `NewTestUpdateStatus().WithManifestRemovedError().WithManifestAddedError()`. 
 
 Luckily, I don't see usage of `NewTestUpdateDisplayStatus(true, true, true, true, true, true, true)` for now it's over design.

wangxiaoxuan273 replied:
Thanks for the comment. Removed the NewTestDisplayStatus functions.

shizhMSFT reviewed (approved):
LGTM

qweeah reviewed (approved):
LGTM
